### PR TITLE
feat(oidc): implement client type public

### PR DESCRIPTION
**Pull request by james-d-elliott**
This implements the public option for clients which allows using Authelia as an OpenID Connect Provider for cli applications and SPA's where the client secret cannot be considered secure.

### CODE REVIEW

**Review by clems4ever — COMMENTED** [review]
From a pure code perspective this PR looks good but I'm wondering how it differs from what's described here: https://www.oauth.com/oauth2-servers/single-page-apps/.
I have never seen this public flag anywhere in the OAuth2 documentation so it seems a bit strange to me. Do you have more info about that?

Is this implementation in Fosite the way to implement the implicit grant flow as described here? https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.2

**Comment by clems4ever** [comment]
> From a pure code perspective this PR looks good but I'm wondering how it differs from what's described here: https://www.oauth.com/oauth2-servers/single-page-apps/.
> I have never seen this public flag anywhere in the OAuth2 documentation so it seems a bit strange to me. Do you have more info about that?
> 
> Is this implementation in Fosite the way to implement the implicit grant flow as described here? https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.2

My bad, I found my answers in https://github.com/ory/fosite/issues/78 and https://datatracker.ietf.org/doc/html/rfc6749#section-2.1.

File: docs/configuration/identity-providers/oidc.md [hunk 1]
```diff
@@ -283,6 +300,23 @@ required: yes
...
+<div markdown="1">
+type: bool
+{: .label .label-config .label-purple }
+default: false
+{: .label .label-config .label-blue }
+required: no
+{: .label .label-config .label-green }
+</div>
+
+The public option enables a client where the end user or a malicious application could reasonably obtain the [client id](#id)
```
Comment (clems4ever) [inline docs/configuration/identity-providers/oidc.md:313]: I would add the following link https://datatracker.ietf.org/doc/html/rfc6749#section-2.1 somewhere around here to help people understand what public refers to in the RFC.

File: docs/configuration/identity-providers/oidc.md [hunk 2]
```diff
@@ -283,6 +300,23 @@ required: yes
...
+and [client secret](#secret). This is particularly useful for SPA's and CLI tools. This option requires (and allows)
+setting the [client secret](#secret) to a blank string.
+
+In addition to the standard redirect URI requirements, the host must be `localhost` or `127.0.0.1`. Alternatively a
+redirect URI of exactly `urn:ietf:wg:oauth:2.0:oob` is valid.
```
Comment (clems4ever) [inline docs/configuration/identity-providers/oidc.md:318]: Can you point to a documentation mentioning this URI? I found some links related to Google using that URI but nothing really described in a RFC.

**Review by clems4ever — CHANGES_REQUESTED** [review]

File: docs/configuration/identity-providers/oidc.md (hunk 2 above)
Comment (james-d-elliott) [reply docs/configuration/identity-providers/oidc.md:318]: Basically I had a look at what Dex did in this area. However on closer inspection they allow any redirect URI for public clients now. I think forcing clients which redirect to `localhost`, `127.0.0.*`, or `urn:ietf:wg:oauth:2.0:oob` to be public is longeterm going to be a reasonable safeguard. However upon inspection we should ensure  the OOB URI is not used with the token response type: https://github.com/dexidp/dex/blob/master/server/oauth2.go#L557-L559

I think what we should do for now is:

- allow public clients which is required to support cli tools safely like argo
- prevent localhost redirect uri's on anything other than public clients
- not support the OOB urn

Thoughts @clems4ever ?

File: docs/configuration/identity-providers/oidc.md (hunk 1 above)
Comment (james-d-elliott) [reply docs/configuration/identity-providers/oidc.md:332]: Added, thanks for the link!

File: docs/configuration/identity-providers/oidc.md (hunk 2 above)
Comment (clems4ever) [reply docs/configuration/identity-providers/oidc.md]: So if fosite already does it, let's trust them imo. It's better that the RFC implementation remains fosite duty altogether.

Comment (clems4ever) [reply docs/configuration/identity-providers/oidc.md]: But we can keep some part of the documentation regarding the fact that only OOB should be used though.

**Review by clems4ever — DISMISSED** [review]
Adapt the doc according to the last comment I left and we are good to go from my point of view.

**Pull request merged**